spec/connection.md

```diff
@@ -0,0 +1,229 @@
+
+#### Max reconnects
+
+**default: 3 / none
+
+Specifies the number of consecutive reconnect attempts the client will make before giving up.
+This is useful for preventing `zombie services` from endlessly reaching the servers, but it can also
+be a footgun and surprise for users who do not expect that the client can give up entirely.
+
+#### Connection timeout
+
+**default 5s**
+
+Specifies how long the client will wait for the TCP connection to be established.
+In some languages, this can hang eternally, and timeout mechanics might be necessary.
+In others, the TCP Connection method might have a way to configure its timeout.
+
+#### Custom reconnect
+
+**Default: none**
+
+If fine-grained control over reconnect attempts intervals is needed, this option allows users to specify one.
+Implementation should make sense in a given language. For example, it can be a callback `fn reconnect(attempt: int) -> Duration`.
+
+#### Tls required
+
+**default: false**
+If set, the client enforces the TLS, whether the Server also requires it or not.
+
+If `tls://` scheme is used in the connection string, this also enforces tls.
+
+#### Ingore advertised servers
+
+**default: false**
+When connecting to the Server, it may send back a list of other servers in the cluster of which it is aware.
+This can be very helpful for discoverability and removes the need for the client to pass all servers in `connect`,
+but it also may be unwanted if, for example, some servers URLs are unreachable for a given client.
+
+#### Retain servers order
+
+**default: false**
+By default, if many server addresses are passed in the connect string or array, the client will try to connect to them in random order.
+This helps healthy connection distribution, but if in a specific case list should be treated as a preference list,
+randomization may be turned off.
+
+### Protocol Commands and Grammar
+
+#### INFO
+[LINK][LINK]
+
+Send by the Server before or after establishing TLS, depending of flow used.
+It contains information about the Server, the nonce, and other server URLs to which the client can connect.
+
+#### CONNECT
+[CONNECT][CONNECT]
+
+Send by the client in response to INFO.
+Contains information about client, including optional signature, client version and connection options.
+
+
+#### Ping Pong
+This is a mechanism to detect broken connections that may not be reported by a TCP connection in a given language.
+
+If the Server sends `PING`, the client should answer with `PONG`.
+If the Client sends `PING`, the Server should answer with `PONG`.
+
+If two consecutive `PONGs are missed, the client should treat the connection as broken, and it should start reconnect attempts.
+
+The default interval for PING is 2 minutes.
+
+#### Client connection related options
+These options should be available to end users, allowing control over connection handling.
+
+#####
+
+### Error Handling (TODO)
+
+Server can respond with `Authorization Error`.
```

Comment:
discuss the specific way in which this is handled - if it terminates retries or not